l3kernel/l3tl.dtx

```diff
@@ -3131,10 +3220,10 @@
 %    \begin{macrocode}
 \prg_new_conditional:Npnn \tl_if_head_is_group:n #1 { p , T , F , TF }
   {
-    \if_catcode:w
+    \if:w
         \exp_after:wN \use_none:n
           \exp_after:wN { \exp_after:wN { \token_to_str:N #1 ? } }
-        * *
+        \scan_stop: \scan_stop:
       \prg_return_false:
     \else:
       \prg_return_true:
```

Comment:
Conditionals are optimized when they end with `\prg_return_true: \else: \prg_return_false: \fi:` but not in the opposite order.  Here I think doing `\reverse_if:N \if:w ... \prg_return_true: \else: \prg_return_false: \fi:` would be faster despite the small overhead of `\reverse_if:N`.  This change may also be useful in a few other places, but I know I already tried to make many conditionals end in the right way to get the optimization.  Seems like I missed this one.

Reply:
Alternative idea: We change `\prg_new_conditional` to also optimize in this case.

Reply:
I don't like that very much, as it would require quite a bit more code than just putting a `\reverse_if:N`.  It's not hard, but to optimize this case in `\prg_new_conditional` we would need to implement the test for presence of `\prg_return_false: \else: \prg_return_true: \fi:` and implement specific code for the `p`, `T`, `F`, `TF` cases.

Reply:
Using `\reverse_if:N` wouldn't work correctly, that'd keep a stray `\scan_stop:` in the input stream (same issue as with `\@@_if_head_is_space:w` below), so the only ways to further optimise this is either by changing `\prg_new_conditional` or by defining the `p`, `TF`, `T`, and `F` variants by hand here.

Reply:
I'd really prefer changing `\prg_new_conditional` (and while at it, change it to not need `\exp_after:wN` but instead let it input something like `\use_i_iii:nnnn \fi: \use_ii:nn` (or, being even faster, some macro defined as `\cs_new:Npn \__prg_fast_fi_i:w \fi: \use_ii:nn #1 #2 { \fi: #1 }`).

Reply:
I would really like not to lengthen `\prg_new_conditional` just to optimize one or two conditionals.  One way out is to replace `\prg_return_false: \else: \prg_return_true: \fi:` by `\exp_after:wN \reverse_if:N \fi: \if_true: \prg_return_true: \else: \prg_return_false: \fi:` in this particular code.

On the other hand, you may be right that the precise code used in `\prg_new_conditional` could be optimized a bit.